py/mpconfig.h

```diff
@@ -1503,6 +1508,19 @@ typedef double mp_float_t;
 #define MP_UNLIKELY(x) __builtin_expect((x), 0)
 #endif
 
+// This statement and onwards are unreachable (helps the compiler eliminate dead code when
+// it can't deduce so automatically)
+#ifndef MP_UNREACHABLE
+#if defined(__GNUC__)
+#define MP_UNREACHABLE() __builtin_unreachable()
+#else
+STATIC NORETURN inline void MP_UNREACHABLE(void) {
+    for (;;);
```

Comment:
This is not really unreachable: it's reachable and will loop forever!  `nlr_jump` had such an implementation only because it must not return.  I'd prefer not to have such an abstraction like this if it's not really needed.

Reply:
So in the non-`__GNUC__` case we want it to do nothing?

I'd say we want to hang *whenever* we reach an`MP_UNREACHABLE()` because that really shouldn't happen.

Reply:
Removed (https://github.com/micropython/micropython/pull/4727#discussion_r280681053)